lms/djangoapps/teams/views.py

```diff
@@ -0,0 +1,449 @@
+    authentication_classes = (SessionAuthentication, OAuth2Authentication)
+    permission_classes = (permissions.IsAuthenticated, IsActiveOrReadOnly)
+
+    paginate_by = 10
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = CourseTeamSerializer
+
+    def get_serializer_context(self):
+        """
+        Adds expand information from query parameters to the serializer context to support expandable fields.
+        """
+        result = super(TeamsListView, self).get_serializer_context()
+        result['expand'] = [x for x in self.request.QUERY_PARAMS.get('expand', '').split(',') if x]
+        return result
+
+    def get(self, request):
+        """
+        GET /api/team/v0/teams/
+        """
+        result_filter = {
+            'is_active': True
+        }
+
+        if 'course_id' in request.QUERY_PARAMS:
+            try:
+                course_key = CourseKey.from_string(request.QUERY_PARAMS['course_id'])
+                result_filter.update({'course_id': course_key})
+            except InvalidKeyError:
+                return Response({'detail': "course_id is not valid"}, status=status.HTTP_400_BAD_REQUEST)
+        if 'topic_id' in request.QUERY_PARAMS:
+            result_filter.update({'topic_id': request.QUERY_PARAMS['topic_id']})
+        if 'include_inactive' in request.QUERY_PARAMS and request.QUERY_PARAMS['include_inactive'].lower() == 'true':
+            del result_filter['is_active']
+        if 'text_search' in request.QUERY_PARAMS:
+            return Response({'detail': "text_search is not yet supported"}, status=status.HTTP_400_BAD_REQUEST)
+
+        queryset = CourseTeam.objects.filter(**result_filter)
+        queryset = queryset.extra(select={'lower_name': "lower(name)"})
+
+        order_by_field = 'lower_name'
+        if 'order_by' in request.QUERY_PARAMS:
+            order_by_input = request.QUERY_PARAMS['order_by']
+            if order_by_input == 'name':
+                order_by_field = 'lower_name'
+            elif order_by_input == 'open_slots':
+                queryset = queryset.annotate(team_size=Count('users'))
+                order_by_field = 'team_size'
+            elif order_by_input == 'last_activity':
+                return Response({'detail': "last_activity is not yet supported"}, status=status.HTTP_400_BAD_REQUEST)
+
+        queryset = queryset.order_by(order_by_field)
+
+        if not queryset:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        page = self.paginate_queryset(queryset)
+        serializer = self.get_pagination_serializer(page)
+        return Response(serializer.data)  # pylint: disable=maybe-no-member
+
+    def post(self, request):
+        """
+        POST /api/team/v0/teams/
+        """
+        field_errors = {}
+        course_key = None
+
+        try:
+            course_key = CourseKey.from_string(request.DATA.get('course_id'))
```

Comment:
I struggled with that when writing the spec. We decided that teams were global in scope, and that a reference to a course was optional. Enforcing course enrollment seems to break that separation that teams are an independently managed resource. However, I can see that this would then leave the API open to malicious updates (create a team in a course you are not enrolled in, switch the course for a team etc).

Given this, I now think that we should enforce course restrictions for non-staff users. This would mean that they can only create teams in courses that they are enrolled in, and that they cannot switch the course for a team. In fact, the latter might already be covered because only staff can update teams at the moment (although I think that restriction will need to be relaxed at some point).